pkg/cmd/codespace/codespace_selector.go

```diff
@@ -0,0 +1,138 @@
+package codespace
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/cli/cli/v2/internal/codespaces/api"
+	"github.com/spf13/cobra"
+)
+
+type CodespaceSelector interface {
```

Comment:
@samcoe an additional struggle I had was method visability - technically anyone in the codespace package can reference `codespaceSelector` (lowercase `c`) directly and use the helpers, so I made the interface to avoid that, but the abstraction did leak in a few places. I tried to create a new package instead, but there was so much code untangling to do and cyclic imports that it wasn't very tenable.

Reply:
Personally I find it more readable to have the "public" well defined, but agree that it does look a little messy to accomplish it.

That said, the other reason I made the interface was to support test injection, which was one of the bonuses you had mentioned 😄

Reply:
I was thinking more about the test injection and as the `apiClient` is already an interface with a generated mock I thought perhaps it makes the most sense to do the injection at that level? Using something like this when you need a selector in testing: 

```go
testSelector := codespaceSelector{
    api: &apiClientMock{
        GetCodespaceFunc: ...
        ...
    },
    repoName: "cli/cli",
    codespaceName: "cs-name-1",
}
```

Reply:
That is a good point that the tests would need to be a bit more intelligent/understanding of the selector functionality. I think the selector can work as a concrete type or an interface, whichever one you prefer. For me I don't like to introduce interfaces unless there are multiple implementations of it, excluding tests, but sometimes that is a high barrier and can make testing a bit more challenging.

Reply:
Thinking about it more, the selector is (currently at least) easy enough to force a selection just by setting `selector.codespaceName` so it's not too burdensome to make it act as a mock. I removed the interface.